feat: add arc-length reparametrization of parametrized curves

michael-novak-math opened the pull request:
add new definitions of arc-length reparametrization and its corresponding parameter transformation and a theorem establishing the desired properties.

michael-novak-math committed:
Changes to be committed:
	new file:   Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean
	modified:   docs/references.bib

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

michael-novak-math committed:
Merge branch 'master' into arclength-reparametrization-branch

michael-novak-math committed:
Changes to be committed:
	modified:   Mathlib.lean

michael-novak-math committed:
Merge branch 'arclength-reparametrization-branch' of https://github.com/michael-novak-math/mathlib4 into arclength-reparametrization-branch
fixing versions

grunweg commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,291 @@
+## References
+ - We roughly followed [Christian_Bar_2010].
+-/
+
```
You also want to add this:
```suggestion

@[expose] public section

```

grunweg commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,291 @@
+import Mathlib.Analysis.Calculus.IteratedDeriv.Defs
+import Mathlib.MeasureTheory.Integral.IntervalIntegral.FundThmCalculus
+import Mathlib.Order.BourbakiWitt
+
```
I'm also surprised to see the `BourbakiWitt` import; my intuition is that it surely can be replaced. But we can revisit this later, once all other kinds in your code are ironed out.

grunweg commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,291 @@
+
+variable {r : WithTop ℕ∞} {t₀ : ℝ}
+
+--Auxilary definition, so we don't have to type the full long expression everytime
```
```suggestion
-- Auxilary definition, so we don't have to type the full long expression every time
```

grunweg reviewed (commented):
Thanks for your PR!

I took a quick look at your PR. You have two CI errors at the moment. The first set of errors is a linter telling you about some small code style issues to fix. Can you fix those, please? It mostly tells you what to do already.
The second error ("cannot import non-module inside a module") I've told you how to fix. I'm happy to explain this more in mid-January, but people on zulip also will be happy to!

While I was looking at it, I left a few comments on code style. This is not a full review, but I figured some comments are better than none.

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: Michael Rothgang <rothgang@math.uni-bonn.de>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: Michael Rothgang <rothgang@math.uni-bonn.de>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: Michael Rothgang <rothgang@math.uni-bonn.de>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: Michael Rothgang <rothgang@math.uni-bonn.de>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: Michael Rothgang <rothgang@math.uni-bonn.de>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: Michael Rothgang <rothgang@math.uni-bonn.de>

michael-novak-math committed:
Update ArcLengthReparametrization.lean

removed a theorem that needs to go somewhere else

michael-novak-math committed:
Update FundThmCalculus.lean

add a version of the first part of the fundamental theorem of calculus for a continuous function on a general interval

michael-novak-math commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,291 @@
+-/
+
+/-- This is a version of the first part of FTC for a function which is continuous on an interval.
+    Credit to Aaron Liu. -/
```
I moved it now. Hopefully this is in the right place now.

grunweg commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,291 @@
+import Mathlib.Analysis.Calculus.InverseFunctionTheorem.Deriv
+import Mathlib.Analysis.Calculus.IteratedDeriv.Defs
+import Mathlib.MeasureTheory.Integral.IntervalIntegral.FundThmCalculus
+import Mathlib.Order.BourbakiWitt
```
I just checked out your PR and cannot reproduce this. Perhaps asking on zulip will help.

grunweg commented:
By the way: once you have addressed all comments and CI passes, you can post a comment with just the text `awaiting-author`, and the corresponding label will be removed. This marks your PR as ready for review again. (Otherwise, it will usually not be looked at.)

michael-novak-math committed:
Update FundThmCalculus.lean

change the new theorem name

michael-novak-math committed:
Update ArcLengthReparametrization.lean

change the name of a theorem used in a proof after its name was changed

michael-novak-math commented:
awaiting-author

grunweg commented:
You still see a red cross next to the "Build (fork)" check on this PR. In the build step there, you still find warnings. (You should also see these warnings if you open the file in VS Code!) You have fixed some warnings, but not those.

michael-novak-math committed:
Update ArcLengthReparametrization.lean

some style fixes

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

michael-novak-math committed:
Update ArcLengthReparametrization.lean

more style fixes

michael-novak-math committed:
Update ArcLengthReparametrization.lean

fixed a wrong name

michael-novak-math committed:
Update ArcLengthReparametrization.lean

omitting some unnecessary assumptions

michael-novak-math committed:
Small syntax errors fixed

grunweg commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,266 @@
+/-- Auxilary lemma - ψ is continuous. -/
+lemma revParamTransform_continuousOn_aux (once_diff : 1 ≤ r) (ht₀ : t₀ ∈ I)
+  (hc : ContDiffOn ℝ r c I) (hI : IsOpen I) : ContinuousOn ψ I :=
+  fun t ↦ fun ht ↦ (revParamTransform_deriv_eq_aux c I ht once_diff ht₀ hc hI).continuousWithinAt
```
```suggestion
  fun _t ht ↦ (revParamTransform_deriv_eq_aux c I ht once_diff ht₀ hc hI).continuousWithinAt
```
This should also fix the unused variable warning.

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: Michael Rothgang <rothgang@math.uni-bonn.de>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: Michael Rothgang <rothgang@math.uni-bonn.de>

michael-novak-math committed:
join multiple rw lines into one

michael-novak-math commented:
-awaiting-author

michael-novak-math committed:
Another two rw lines joined into one.

michael-novak-math committed:
Small English mistake fix.

SnirBroshi commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,262 @@
+
+variable {r : WithTop ℕ∞} {t₀ : ℝ}
+
+-- Auxilary definition, so we don't have to type the full long expression every time
```
```suggestion
-- Auxiliary definition, so we don't have to type the full long expression every time
```

SnirBroshi commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,262 @@
+  exact intervalIntegral.hasDerivWithinAt_of_continuousOn_interval
+    (speed_continuousOn c I hI hc once_diff) ht₀ ht
+
+/-- Auxilary lemma - ψ is continuous. -/
```
```suggestion
/-- Auxiliary lemma - ψ is continuous. -/
```

SnirBroshi commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,262 @@
+  · apply norm_pos_iff.mpr (regular t ht)
+  · apply hI.uniqueDiffWithinAt ht
+
+/-- Auxilary lemma - ψ is injective. -/
```
```suggestion
/-- Auxiliary lemma - ψ is injective. -/
```

SnirBroshi commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,262 @@
+  have h : StrictMonoOn ψ I := strictMonoOn_of_hasDerivWithinAt_pos hI hψ hψ' hψ'₀
+  h.injOn
+
+/-- Auxilary lemma - ψ is bijective. -/
```
```suggestion
/-- Auxiliary lemma - ψ is bijective. -/
```

SnirBroshi commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,262 @@
+  apply Function.invFunOn_eq
+  exact (Set.mem_image ψ I s).mp hs
+
+--Auxilary definition, so we don't have to type the full long expression every time
```
```suggestion
/-- Auxiliary definition, so we don't have to type the full long expression every time -/
```

SnirBroshi commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,262 @@
+  rw [hinv] at hfinal
+  exact hfinal
+
+/-- This is the main theorem: arclengthParamTransform is an orientation-preserving parameter
```
```suggestion
/-- This is the main theorem: `arclengthParamTransform` is an orientation-preserving parameter
```

SnirBroshi commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,262 @@
+
+/-- This is the main theorem: arclengthParamTransform is an orientation-preserving parameter
+transformation, that is: a bijection from some interval J to the given interval I with a positive
+derivative; and the resulting reparametrization, i.e arclengthReparam is indeed a
```
```suggestion
derivative; and the resulting reparametrization, i.e `arclengthReparam` is indeed a
```

SnirBroshi reviewed (commented):
Hello 👋 
Just some typos that I noticed

I suggest the [Code Spell Checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) VSCode extension to check the whole file

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: SnirBroshi <26556598+SnirBroshi@users.noreply.github.com>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: SnirBroshi <26556598+SnirBroshi@users.noreply.github.com>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: SnirBroshi <26556598+SnirBroshi@users.noreply.github.com>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: SnirBroshi <26556598+SnirBroshi@users.noreply.github.com>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: SnirBroshi <26556598+SnirBroshi@users.noreply.github.com>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: SnirBroshi <26556598+SnirBroshi@users.noreply.github.com>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: SnirBroshi <26556598+SnirBroshi@users.noreply.github.com>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: SnirBroshi <26556598+SnirBroshi@users.noreply.github.com>

michael-novak-math committed:
Update Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean

Co-authored-by: SnirBroshi <26556598+SnirBroshi@users.noreply.github.com>

SnirBroshi commented on lake-manifest.json:
```diff
@@ -65,7 +65,7 @@
    "subDir": null,
    "scope": "leanprover-community",
-   "rev": "dff865b7ee7011518d59abfc101c368293173150",
+   "rev": "2e16f91af2a97975e5d2fac906494cd6c17ba255",
```
Is this necessary? This updates Batteries by a single [commit](https://github.com/leanprover-community/batteries/commit/2e16f91af2a97975e5d2fac906494cd6c17ba255) which doesn't seem related

michael-novak-math committed:
Changing spelling from "arclength" to "arc-length" in comments and documentation.

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

michael-novak-math committed:
Spelling

michael-novak-math committed:
Merge branch 'arclength-reparametrization-branch' of https://github.com/michael-novak-math/mathlib4 into arclength-reparametrization-branch

spelling

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

j-loreaux commented on Mathlib/Analysis/Calculus/ArcLengthReparametrization.lean:
```diff
@@ -0,0 +1,263 @@
+
+/-- Auxiliary definition, this is the reversed (inverse) parameter transformation used for
+constructing the arc-length reparametrization of parametrized curve. -/
+def arclengthParamTransformAux (t₀ : ℝ) : ℝ → ℝ := fun t ↦ ∫ τ in t₀..t, ‖deriv c τ‖
```
I think this is *not* the right generality for arc length reparameterization. In particular, there is an open PR (#31766) defining arc length for general curves in a pseudo metric space (I've just now realized after looking that there have been replies to my reviews but the `awaiting-author` tag was never removed, so I will need to review it again soon). Once we have that, then that function would take the place of this one, and it would be defined for curves that are (a) in metric spaces, not just normed spaces, (b) not necessarily differentiable. Then for any rectifiable (or even locally rectifiable) curve, we can have an arc length reparameterization. It could then be shown that this coincides with the derivative approach when the curve satisfies sufficient differentiability hypotheses.

j-loreaux replied:
I think it odd to claim that you:

> didn't define in this file the arc-length

when the code I'm highlighting in this comment:
```lean
def arclengthParamTransformAux (t₀ : ℝ) : ℝ → ℝ := fun t ↦ ∫ τ in t₀..t, ‖deriv c τ‖
```
is literally a definition of arc-length for sufficiently nice curves (https://en.wikipedia.org/wiki/Curve#Length_of_a_curve)

You define the arc-length and then use it to define the arc-length reparameterization. But my point is that none of it is general enough, and we should have this for general curves. Then curves which are (locally) rectifiable will have an arc-length reparameterization.

> If you also plan to define arc-length reparametrization to general curves, do you think it would be possible to show that they have unit-speed for regular parametrized curves and so show that they are the same as what I defined?

I answered this already:

> It could then be shown that this coincides with the derivative approach when the curve satisfies sufficient differentiability hypotheses.

michael-novak-math replied:
> I think it odd to claim that you:
> 
> > didn't define in this file the arc-length
> 
> when the code I'm highlighting in this comment:
> 
> ```lean
> def arclengthParamTransformAux (t₀ : ℝ) : ℝ → ℝ := fun t ↦ ∫ τ in t₀..t, ‖deriv c τ‖
> ```

It's very similar, but this is a parameter transformation, not the length. I don't mind defining my parameter transformation using your general definition for arc-length once we also have the the theorem that it's the same as the derivative approach, but this is a small part of my file, absolutely not the main point of my file. The main point of my file is to show that every regular parametrized curve has a unit-speed reparametrization (also known as arc-length reparametrization). So I think that besides this one line which we can change to the general defintion of arc-length once we have it in mathlib (and the theorem that it's equal to what I've written here) all the rest of the file is still new contribution that that doesn't make sense in more general context, because for a curve to have unit-speed it needs to be differentiable.

j-loreaux reviewed (commented):
@grunweg do you agree with my assessment?

michael-novak-math committed:
Merge branch 'master' into arclength-reparametrization-branch

michael-novak-math committed:
Merge branch 'master' into arclength-reparametrization-branch

michael-novak-math committed:
Merge branch 'master' into arclength-reparametrization-branch

michael-novak-math committed:
Merge branch 'master' into arclength-reparametrization-branch

michael-novak-math committed:
Merge branch 'master' into arclength-reparametrization-branch

grunweg commented:
Let me return to this PR: I think we want to have this concept in mathlib, but we should do it properly.
- there should be a definition of the arc-length of curves first, that works for any rectifiable curve (not just differentiable ones)
- then, you should show that the arc-length of a C¹ curve coincides with the formula you gave;
- finally, you can take in the remainder of your PR to define the arc-length reparametrisation.

In particular, parts of this PR are definitely useful, the initial pieces just need to be done differently.
